bin/public_interface.py

```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env python
+"""Extract/compare public interfaces."""
+import argparse
+import importlib
+import inspect
+import json
+import os.path
+import pkgutil
+from pathlib import Path
+from typing import Any, Dict, Set, Union
+
+
+class InterfaceScanner:
+    def __init__(self) -> None:
+        self.signatures: Dict[str, Union[inspect.Signature]] = {}
+        self.constants: Set[str] = set()
+
+    def scan_interfaces_recursively(self, module_name: str) -> None:
+        self._scan_interfaces_in_module(module_name)
+        for submodule in pkgutil.iter_modules([module_name.replace(".", os.path.sep)]):
+            submodule_name = module_name + "." + submodule.name
+            self.scan_interfaces_recursively(submodule_name)
+
+    def _scan_interfaces_in_module(self, module_name: str) -> None:
+        self._scan_functions_in_module(module_name)
+        self._scan_classes_in_module(module_name)
+        self._scan_constants_in_module(module_name)
+
+    def _scan_functions_in_module(self, module_name: str) -> None:
+        for function_name, function in inspect.getmembers(
+            importlib.import_module(module_name), lambda obj: inspect.ismethod(obj) or inspect.isfunction(obj)
+        ):
+            # Skip imported functions and ones starting with "_"
+            if function.__module__ != module_name or function_name.startswith("_"):
+                continue
+            full_path = f"{module_name}.{function_name}"
+            self.signatures[full_path] = inspect.signature(function)
+
+    def _scan_constants_in_module(self, module_name: str) -> None:
```

Comment:
It's not necessarily constants though. Nor do I think it matters. These are module-scoped variables, we don't know if they mutate. Maybe just `variables`?

Reply:
updated